add a warp by vector filter to filters

flothesof opened the pull request:
### Overview

Following discussion in #650 and encouragements of @imsodin, I'm proposing a suggestion for a warp by vector filter, inspired by the conceptually close warp_by_scalar filter.

### Details

- adds a warp_by_vector filter in pyvista/core/filters.py

I would need some additional input regarding two things:
- how to add error checking at the beginning of function
  - should it check for empty argument list after popping factor from kwargs?
  - should it raise an error if there are no point data fields?
  - is it right to name the argument of the function scalars if they're vectors?
- how to design some possible tests
  - I'm thinking of a 3D mesh and adding simple vector fields like translations and rotations where we could check the warped coordinates
- adding an example that looks nice (like eigenmodes of a plate or something like that)

Anyways, thanks for the amazing library. I really enjoy using pyvista.

Regards,
Florian

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -1129,6 +1129,24 @@ def warp_by_scalar(dataset, scalars=None, factor=1.0, normal=None,
             return
         return output
 
+    def warp_by_vector(dataset, scalars=None, factor=1.0):
+        """Warp the dataset's points by a point data vectors array's values.
+
+        This modifies point coordinates by moving points along point vectors by
+        the local vector times the scale factor.
+
+        A classical application of this transform is to visualize eigenmodes in mechanics.
+        """
```
Please expand the docstring to include `Parameters`, `Returns`, and `Example`.  See the other docstrings for an example.

One note, anything you write in the `Examples` section will be run within the pipeline and have its outputs checked against some expected value (anything after `>>>`).  Make sure if something returns something, it's set to a variable so it doesn't automatically print out.

flothesof replied:
It would be great to have a small example, but I'm not sure what would work here. Do you have any pointers to a simple to setup example which could be applied to the vector warping?

akaszynski commented:
Thanks for adding the PR!  It's always great to have new contributors from the community and userbase.

> how to add error checking at the beginning of function
You could leave it alone, or for bonus points use `isinstance` and wrap non `pyvista` datasets.  Look for other methods/functions that use `pyvista.wrap`.

> should it check for empty argument list after popping factor from kwargs?
Since you're not using `**kwargs`, you won't have to worry about this.  Additional keyword arguments will raise an error

> should it raise an error if there are no point data fields?
Yes.

> is it right to name the argument of the function scalars if they're vectors?
Vectors does make more sense, and vectors could also mean the name of the point array you'd like to grab from the `dataset`.  So, `array_name` or `array` might work as well.

> how to design some possible tests
> I'm thinking of a 3D mesh and adding simple vector fields like translations and rotations where we could check the warped coordinates
That would work.  Make it something simple with a relatively known output.  Compute the translation and rotation using some other package (`numpy`) and then see if you get the same results here.  That way if we break something we'll know.

> adding an example that looks nice (like eigenmodes of a plate or something like that)
That would be great!

akaszynski commented on pyvista/core/filters.py:
```diff
@@ -1129,6 +1129,24 @@ def warp_by_scalar(dataset, scalars=None, factor=1.0, normal=None,
+        alg.SetScaleFactor(factor)
+        alg.Update()
+        output = _get_output(alg)
+        return output
```
Consider updating "in-place".  You won't have to return an entirely new mesh.  Look at other methods/functions that use the `in_place` keyword argument for an example.

akaszynski commented:
As for the tests, don't worry if the greeting fails.  I have to look into why that's failing.

banesullivan commented on pyvista/core/filters.py:
```diff
@@ -1129,6 +1129,24 @@ def warp_by_scalar(dataset, scalars=None, factor=1.0, normal=None,
             return
         return output
 
+    def warp_by_vector(dataset, scalars=None, factor=1.0):
```
the argument should be `vectors` instead of `scalars` because that's what is required for warping

flothesof commented on pyvista/core/filters.py:
```diff
@@ -1091,10 +1091,10 @@ def warp_by_scalar(dataset, scalars=None, factor=1.0, normal=None,
         scalars : str, optional
-            Name of scalars to warb by. Defaults to currently active scalars.
+            Name of scalars to warp by. Defaults to currently active scalars.
 
         factor : float, optional
-            A scalaing factor to increase the scaling effect. Alias
+            A scaling factor to increase the scaling effect. Alias
```
I've also spotted some typos in adjacent docstrings, so here's a suggestion for fixing them.

banesullivan replied:
thanks!

flothesof commented:
I've updated the PR following @akaszynski and @banesullivan helpful code-review.

flothesof commented:
I've now also added a proposed example for eigenmode computation and warping.
![image](https://user-images.githubusercontent.com/2360669/78769203-4bf63e00-798d-11ea-9cc7-238067c5cf85.png)

flothesof commented:
> That example is pretty crazy! Thanks for all the effort! It appears all the options of the filter are exposed so LGTM

Thanks for giving me the opportunity to share :)

> One small request: I would like to see some simple tests implemented for this filter just to make sure it is working. Simply pass a mesh and check the output type and that the points are not equal. We don't want to actually test the result as VTK should be doing that on their side

Regarding the tests, there's two possibilities in my opinion:

- adding a simple call on one of the examples of pyvista
- a slightly more involved one would be to replicate the warp by scalar elevation map example using warp by vector with a vector field pointing upwards (but this option is probably not the motss relevant since as you mentioned this is really for vtk to do this type of check)

banesullivan commented:
Hm. Give me one minute, I am going to re-review this.

banesullivan commented on pyvista/core/filters.py:
```diff
@@ -1129,6 +1129,41 @@ def warp_by_scalar(dataset, scalars=None, factor=1.0, normal=None,
+
+        """
+        if vectors is None:
+            field, vectors = dataset.active_scalars_info
```
make sure that the filter is capable of leveraging vectors that are set under the `vectors` attribute (i.e. have the name `"_vectors"`). This should be as simple as changing this line to 

```suggestion
            field, vectors = dataset.active_vectors_info
```

And then also be sure to check that the array is indeed a vector array (has three components), throwing an error if not.

banesullivan commented:
Note that I just adjusted that code

flothesof commented:
Thanks for the suggestions. 
I've included your proposed sphere to the examples, added a simple example with it and moved the eigenmodes example to the advanced section.

Let me know what you think.

Regards,
Florian

flothesof commented on pyvista/core/filters.py:
```diff
@@ -1156,7 +1156,7 @@ def warp_by_vector(dataset, vectors=None, factor=1.0, inplace=False):
             field, vectors = dataset.active_vectors_info
         arr, field = get_array(dataset, vectors, preference='point', info=True)
         # check that this is indeed a vector field
-        if field != pyvista.POINT_DATA_FIELD or arr.ndim != 2 or arr.shape[1] != 3 :
```
This is weird: I've copied that part from warp_by_scalar but it failed the tests (maybe this is not tested in warp_by_scalar?). I'm therefore removing it.

banesullivan commented on pyvista/core/filters.py:
```diff
@@ -1129,6 +1129,46 @@ def warp_by_scalar(dataset, scalars=None, factor=1.0, normal=None,
+                                 'The values you provided do not satisfy this requirement')
+        alg = vtk.vtkWarpVector()
+        alg.SetInputDataObject(dataset)
+        alg.SetInputArrayToProcess(0, 0, 0, field, vectors)
```
```suggestion
        alg.SetInputArrayToProcess(0, 0, 0, field.value, vectors)
```

The field was changed to an enum type recently on master

flothesof commented:
Note: I rebased the changes on the latest master (which explains the enum stuff failing in the CI but not in my local state).

banesullivan commented:
Thanks for all the hard work @flothesof! The changes/tests look great to me. The only thing left at this point is that some lines are too long. We typically only have line lengths of 79 characters

flothesof commented:
If I may complain about your comment: if you really want to enforce the line lengths, shouldn't this be handled by the CI? I've noticed that core\filters.py doesn't strictly enforce 79 character line lengths...

Anyhow, what do you think about the updated files?

banesullivan commented:
Everything looks great now! the line lengths are mostly a guideline.. there's a discussion about it somewhere in some issue and we decided not to have it on the CI (yet). Mostly because some places would just get ugly trying to always adhere to that line length and the CIs complain enough already. One day, we'll just set up an auto-formatter, but not anytime soon as we have way too much going on.

Thanks for changing those lines - I'll get this merged before long!
